libraries/AP_Compass/AP_Compass_Backend.cpp

```diff
@@ -95,6 +97,57 @@ void AP_Compass_Backend::correct_field(Vector3f &mag, uint8_t i)
     mag += state.motor_offset;
 }
 
+void AP_Compass_Backend::accumulate_sample(Vector3f &field, uint8_t instance,
+                                           uint32_t max_samples)
+{
+    /* rotate raw_field from sensor frame to body frame */
+    rotate_field(field, instance);
+
+    /* publish raw_field (uncorrected point sample) for calibration use */
+    publish_raw_field(field, instance);
+
+    /* correct raw_field for known errors */
+    correct_field(field, instance);
+
+    if (!field_ok(field)) {
+        return;
+    }
+
+    WITH_SEMAPHORE(_sem);
+
+    _accum += field;
+    _accum_count++;
+    if (max_samples && _accum_count >= max_samples) {
+        _accum_count /= 2;
+        _accum /= 2;
+    }
+}
+
+void AP_Compass_Backend::drain_accumulated_samples(uint8_t instance,
```

Comment:
The implementation here could do with looking at.  Some of the equivalent code you've removed did better in terms of narrowing the time the semaphore is held than others.  This is no worse than a lot of them, but worse than some of them :-)  This basically comes down to taking a local copy of `_accum / _accum_count`.  In particular, publishing the field while holding the semaphore seems unfortunate.

Still - now we have *one* place to fix things rather than half-a-dozen!

Reply:
"publish" is actually just copying the value. The additional time holding the semaphore seemed neglectible compared to what's being done.

As you noticed, this can be micro-optimized in one place now